src/error.rs

```diff
@@ -0,0 +1,130 @@
+//! Specs error module.
+//!
+//! There are specific types for errors (e.g. `WrongGeneration`)
+//! and additionally one `Error` type that can represent them all.
+//! Each error in this module has an `Into<Error>` implementation.
+
+use std::error::Error as StdError;
+use std::fmt::{Debug, Display, Formatter, Result as FmtResult};
+
+use world::{Entity, Generation};
+
+/// A boxed error implementing `Debug`, `Display` and `Error`.
+pub struct BoxedErr(pub Box<StdError + Send + Sync + 'static>);
+
+impl BoxedErr {
+    /// Creates a new boxed error.
+    pub fn new<T>(err: T) -> Self
+    where
+        T: StdError + Send + Sync + 'static,
+    {
+        BoxedErr(Box::new(err))
+    }
+}
+
+impl AsRef<StdError> for BoxedErr {
+    fn as_ref(&self) -> &(StdError + 'static) {
+        self.0.as_ref()
+    }
+}
+
+impl Debug for BoxedErr {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "{:}", self.0)
+    }
+}
+
+impl Display for BoxedErr {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+impl StdError for BoxedErr {
+    fn description(&self) -> &str {
+        self.as_ref().description()
+    }
+}
+
+/// The Specs error type.
+/// This is an enum which is able to represent
+/// all error types of this library.
+///
+/// Please note that you should not use `__NonExhaustive`,
+/// which is a variant specifically added for extensibility
+/// without breakage.
+#[derive(Debug)]
+pub enum Error {
+    /// A custom, boxed error.
+    Custom(BoxedErr),
+    /// Wrong generation error.
+    WrongGeneration(WrongGeneration),
+
+    #[doc(hidden)]
+    __NonExhaustive,
```

Comment:
What kind of breakage would occur without this?

Reply:
If a client matches against all the variants and we add one, it would cause an error about non-exhaustive matching. By having this additional variant which should never be used, the user is forced to match against `_`. This is a very common pattern (even the standard library uses it, see https://doc.rust-lang.org/src/core/sync/atomic.rs.html#220) and there's an RFC about creating an attribute `#[non_exhaustive]` for just this case.